feat: support async plugin's pre/post

timofei-iatsenko opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #16860 <!-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  | No
| License                  | MIT


Enable support of async pre/post functions in plugins + update tests.  As discussed here #16860


I'm also wondering of adding a `isAsync` flag for pre/post functions arguments so plugins could adjust it's own logic based on that. However i don't know how exactly to shape this signature, my proposition is: 

```ts
  pre?: (this: S, file: File, ctx: {isAsync: boolean}) => void | Promise<void>;
  post?: (this: S, file: File, ctx: {isAsync: boolean}) => void | Promise<void>;
```

Let me know what you think, maybe there is a better place for that.

timofei-iatsenko commented on packages/babel-core/test/fixtures/async/plugin-post/plugin.js:
```diff
@@ -4,11 +4,12 @@ module.exports = function plugin({ types: t }) {
   return {
     async post() {
       await wait(50);
+      this.file.ast.program.body[0].value = "success"
```
checking that promise is actually awaited

timofei-iatsenko commented on packages/babel-core/src/transformation/index.ts:
```diff
@@ -93,22 +94,14 @@ function* transformFile(file: File, pluginPasses: PluginPasses): Handler<void> {
+      if (plugin.pre) {
+        const fn = maybeAsync(
+          plugin.pre,
+          `You appear to be using an async plugin/preset, but Babel has been called synchronously`,
```
Also not sure about error message here. Should that be adjusted to notify that exactly pre/post function is async, or that general error message would be enough?

nicolo-ribaudo replied:
This should be enough -- from a user's point of view, it doesn't matter which part exactly of a plugin is async :)

nicolo-ribaudo reviewed (commented):
This looks good -- can you also ad a test with an async `pre`, which transforms the AST in such a way that makes sure that it's being awaited before running the visitors?

timofei-iatsenko commented:
@nicolo-ribaudo there is one, it reassigns variable in the closure. If pre would not be awaited, visitor will print "failure" instead of "success". 

I also have question regarding typings. Should we update typings of pre/post to reflect that they are supporting Async result? 
I tried to change them, but this caused a lot of TS errors, and i honestly don't know how to solve them properly in this codebase without big changes.

nicolo-ribaudo commented:
> I also have question regarding typings. Should we update typings of pre/post to reflect that they are supporting Async result?
I tried to change them, but this caused a lot of TS errors, and i honestly don't know how to solve them properly in this codebase without big changes.

I tried changing them in `packages/babel-core/src/config/validation/plugins.ts`, and the type failure actually uncovered a bug!

Where we do `chain(inherits.pre, plugin.pre)`, we are calling the two `pre` functions one after the other without awaiting if the first one is async. I think something like this should work (but I didn't test it):
```js
function chainMaybeAsync<Args extends any[]>(
  a: undefined | ((...args: Args) => void | Promise<void>),
  b: undefined | ((...args: Args) => void | Promise<void>),
) {
  if (!a) return b;
  if (!b) return a;

  return function (this: unknown, ...args: Args) {
    const res = a.apply(this, args);
    if (res != null && typeof (res as Promise<void>).then === "function") {
      return (res as Promise<void>).then(() => b.apply(this, args));
    }
    return b.apply(this, args);
  };
}
```

You can add a test with a plugin A that has plugin B in its `inherits`, plugin B's `pre` is async, and plugin A's `pre` must run after it.

timofei-iatsenko commented:
@nicolo-ribaudo i changed types for pre/post and fixed chaining. Also added two more tests for pre/post with inheritance. 

The pipelines are failing for unknown reason. Could you check what's wrong?

nicolo-ribaudo commented on packages/babel-core/test/fixtures/async/plugin-post-chaining/plugin.js:
```diff
@@ -0,0 +1,35 @@
+  return {
+    inherits: pluginC,
+    async post() {
+      await wait(50);
```
Could you make pluginC async and keep pluginB sync? So we have `async-sync-async`, testing both combinations :)

timofei-iatsenko replied:
Done, test for `pre` uses sync-async-async and test for `post` uses async-sync-async.

nicolo-ribaudo reviewed (approved):
It just needed a rebase.

I left a minor comment, but this looks good now!

timofei-iatsenko commented:
@nicolo-ribaudo what about passing `isAsync` to the plugin's authors? It could be passed directly to the pre/post function as a parameter:

```ts
pre?: (this: S, file: File, ctx: {isAsync: boolean}) => void | Promise<void>;
post?: (this: S, file: File, ctx: {isAsync: boolean}) => void | Promise<void>;
```

Or might be as a property on a PluginPass object? 

I'm afraid it would be a bit tricky to implement no-breaking logic for the plugins without that flag.

nicolo-ribaudo commented:
Whops sorry, I missed the question. I would prefer to expose it on `PluginPass`.

timofei-iatsenko commented:
@nicolo-ribaudo i've added a new property `isAsync` on a `BabelPass` object. Corresponding tests for the functionality added.

timofei-iatsenko commented on packages/babel-core/src/transformation/plugin-pass.ts:
```diff
@@ -7,19 +7,32 @@ export default class PluginPass<Options = object> {
-  // The working directory that Babel's programmatic options are loaded
-  // relative to.
+  /**
+   * The working directory that Babel's programmatic options are loaded
+   * relative to.
+   */
   cwd: string;
 
-  // The absolute path of the file being compiled.
+  /** The absolute path of the file being compiled. */
   filename: string | void;
```
With `/**` style of comment it's a jsdoc description which could be used as quick documentation in IDE.

timofei-iatsenko commented on packages/babel-core/src/transformation/plugin-pass.ts:
```diff
@@ -7,19 +7,32 @@ export default class PluginPass<Options = object> {
+    file: File,
+    key?: string | null,
+    options?: Options,
+    isAsync?: boolean,
```
i've added `isAsync` as optional argument to the last position, so it shouldn't be a breaking change to anyone who uses `PluginPass` constructor directly. However, i would like to add it as required argument after the `file`.  Let me know if that change is acceptable.

nicolo-ribaudo replied:
We only export `PluginPass` as a type and not as a constructor (see `packages/babel-core/src/index.ts`), so it should be fine to add it as a required property here.

nicolo-ribaudo replied:
Actually, you can just remove the `?`s from the other arguments since they are always always passed. It's probably a leftover from when we converted the repo from Flow to TS.

timofei-iatsenko commented on packages/babel-core/test/async.js:
```diff
@@ -308,7 +338,7 @@ describe("asynchronicity", () => {
 
   describe("misc", () => {
-    it("unknown preset in config file does not trigget unhandledRejection if caught", async () => {
+    it("unknown preset in config file does not trigger unhandledRejection if caught", async () => {
```
fixed a typo

JLHwung commented on packages/babel-core/src/config/full.ts:
```diff
@@ -488,16 +488,18 @@ function* loadPresetDescriptor(
-      fn.apply(this, args);
+function chainMaybeAsync<Args extends any[], R extends void | Promise<void>>(
+  a: undefined | ((...args: Args) => R),
+  b: undefined | ((...args: Args) => R),
```
Here we assume `a` and `b` share the same return type, however, it is also possible that `a` is sync while `b` is async. I suggest we remove the `R` constraint and type them as `(...args: Args) => void | Promise<void>`. We can probably get rid of the final type cast as well.

For situations where both `a` and `b` are sync, we can add a type overload so that the chained function is still sync.

JLHwung replied:
On the mixed-type cases, for example, if `inherits.pre` is a sync function and `plugin.pre` is an async function, then the merged `pre` will be an async function.

Like you said, every function is assignable to `() => void`, so the type checking can't help very much here. I am fine with leaving it as-is.

timofei-iatsenko commented:
Thanks for your reviews and helps on this PR.
